stream create: Modernize create-stream add-subscribers UI.

showell opened the pull request:
This is for issue #20499 and replaces #21069

timabbott commented:
This UX feels pretty good ignoring the positioning/CSS issues. I think we should remove the parens around the email addresses to match what we have in edit stream, in case that's not on your list.

showell commented:
Recent changes since Tim QA'ed this:
* added the "Add all users" button
* fixed how we show emails
* added node tests

I would also look at https://chat.zulip.org/#narrow/stream/101-design/topic/stream-create.20pills.20.2320499/near/1338352 for discussion of some known quirks that affect stream-edit as well as my new code.

alya commented:
Nice, I think we're getting close!

@timabbott and I thought some more about the "Add all users" button, and we were thinking that a better way to manage offering that option would probably be within the typeahead. We could make it be the case that "everyone" essentially acts like a group of all users. A couple of options for how to advertise it would be:

1. In the placeholder text, e.g. "Add subscribers. Use usergroup name, #streamname or everyone to bulk add subscribers." It's getting pretty long though, so I'd be worried about it in translation. Maybe we could wordsmith to shorten it.
2. In a hint at the bottom of the typeahead, which will probably work better. E.g. something like "Hint: You can add a usergroup or everyone."

When we do this, we should also move the filter to be on the right of the "Add subscribers" line. We should have it visually match the filter in the stream membership editing UI.

showell commented:
All but the last huge commit here are on  #21314, which we should review first.

alya commented:
For the Subscribers table heading, I think we should use the same font as for the other settings tables.

alya commented:
Did you end up deciding that the "Add all users" button won't work well next to the "Choose subscribers" heading?

alya commented:
Otherwise looks good to me!

showell commented:
> 

Yeah, this was sort of a local decision instead of a global decision, as I didn't want to overpower the other headings in Create Stream, and we haven't actually merged the PR here for all the settings stuff.  We can consolidate the headings a bit after the other PR gets merged.  (This is also sorta related to the question of whether we want to go away from the "Who can access this stream?" bold headings.)

showell commented:
> Did you end up deciding that the "Add all users" button won't work well next to the "Choose subscribers" heading?

I tried like five different things, and none of them really looked perfect to me, so I settled on putting it below the input pill but above the list.  Even though it doesn't look fantastic there, it seems like the right functional place for it, and we can always iterate.

timabbott commented:
This feels pretty slick and certainly an improvement over main; didn't notice any bugs playing with it. We should merge this as soon as we get to a good checkpoint on it.  I have a few notes:
* It might be the case that we want to use the phrase "Choose initial subscribers".
* The scroll position experience is a little messy when using the filter widget. My opinion is that the right way to fix this is the multi-panel approach we've discussed, so no action required.
* We could have a line of text that initial subscribers explaining what we're going to do -- e.g. "These users will be notified that they were added to the new stream".
* I'm a bit worried about the prominent "Add all users" button being a bit too encouraging to press (E.g. folks using that rather than thinking about who to add). Seems like not a blocker.  
* I also wonder for the medium term vision if it makes sense to replace that button with a `dropdown_list_widget` that offered various other special groups, like "All moderators" or a stream or whatever as something that might feel more like a fancier GUI complementing the current typeahead thing. Mentioning this in case it affects thinking about where to place to button.

showell commented:
> This feels pretty slick and certainly an improvement over main; didn't notice any bugs playing with it. We should merge this as soon as we get to a good checkpoint on it. I have a few notes: [...]

I agree with all the suggestions from the prior comment, but since it appears that none are blockers, I think this is ready for a final review.

alya commented:
I just played with this, and it's looking good!

> I'm a bit worried about the prominent "Add all users" button being a bit too encouraging to press (E.g. folks using that rather than thinking about who to add). Seems like not a blocker.

It also seems a bit redundant -- I wonder if we should make it say just "Add",  or perhaps "Add all"? We could also style it the same as the "Add" button to make it stand out less.

alya commented:
Another thing I noticed is that we should let the user re-sort the list of users by different columns (both in the create stream UI and the stream members tab). Not a blocker; let me know if it would be helpful to file an issue for this.

alya commented:
> For sorting the columns, I think it's fine to open an issue for it. I'm not 100% sure it's worth the complexity, since the name column is already sorted. Sorting the email column is a no-op for some users on some realms (due to hidden emails), but it's probably fine if it just feels strange. Sorting by id is kinda strange. And then the action column has all the same buttons.
> 
> So we'd need to be a bit thoughtful here.

Yeah, I was thinking we could follow the same pattern we use in the Org settings > Users table. I can open an issue.

timabbott commented:
I tried changing it to "Add all", and while I agree "Add all users" feels a bit redundant, for i18n and other reasons, I think it's best to spell it out.  We could consider changing the question to "Do you want to add everyone?" to avoid the redundancy.

For the button styling, I think we probably want to change the "Add" button to use the same CSS as the "Create" active button style, so my proposal would be that we do that rather than changing the "Add all users" button.  (We also may want to replace the sea-green button style with our purple active button style from the "send" button, but that's probably a bit of an orthogonal mini project that @amanagr might help with).

alya commented:
> We could consider changing the question to "Do you want to add everyone?" to avoid the redundancy.

Sure, let's do that.

> For the button styling, I think we probably want to change the "Add" button to use the same CSS as the "Create" active button style, so my proposal would be that we do that rather than changing the "Add all users" button.

Yes, you're right, that would be consistent with all the other "Add" buttons in our settings. We should change it both here an in the "edit subscribers' UI.

alya commented:
To really be consistent with other settings, the button should probably be on its own line, and be labeled more fully as "Add subscribers". (not a blocker)

alya commented:
I'm not sure the purple button styling would be a good idea; it might be too bright. Anyway, the green buttons are also consistent with other part so the UI (e.g. saving changes to edit topic), and I'm not excited about getting into a button colors project right now.

timabbott commented:
OK, merged with these changes: 
```
diff --git a/static/templates/stream_settings/add_subscribers_form.hbs b/static/templates/stream_settings/add_subscribers_form.hbs
index 3735b93459..2144142e30 100644
--- a/static/templates/stream_settings/add_subscribers_form.hbs
+++ b/static/templates/stream_settings/add_subscribers_form.hbs
@@ -6,7 +6,7 @@
         </div>
     </div>
     <div class="add_subscriber_btn_wrapper inline-block">
-        <button type="submit" name="add_subscriber" class="button add-subscriber-button small rounded" tabindex="0">
+        <button type="submit" name="add_subscriber" class="button add-subscriber-button small rounded sea-green" tabindex="0">
             {{t 'Add' }}
         </button>
     </div>
diff --git a/static/templates/stream_settings/new_stream_users.hbs b/static/templates/stream_settings/new_stream_users.hbs
index ca652686a9..283fde96aa 100644
--- a/static/templates/stream_settings/new_stream_users.hbs
+++ b/static/templates/stream_settings/new_stream_users.hbs
@@ -3,7 +3,7 @@
 </div>
 <br />
 
-{{t "Do you want to add all users?"}}
+{{t "Do you want to add everyone?"}}
 <button class="add_all_users_to_stream small button rounded sea-green">{{t 'Add all users'}}</button>
 
 <div class="create_stream_subscriber_list_header">
```

(The color change is in a shared component, so only needed to do it in one place).  Thanks @showell for making this happen!
